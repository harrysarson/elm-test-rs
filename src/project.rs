use anyhow::Context;
use notify::watcher;
use notify::RecursiveMode;
use notify::Watcher;
use pubgrub_dependency_provider_elm::project_config::ProjectConfig;
use std::collections::BTreeSet;
use std::iter;
use std::path::PathBuf;
use std::sync::mpsc::channel;
use std::time::Duration;

#[derive(Debug)]
pub struct Project {
    pub config: ProjectConfig,
    pub src_and_test_dirs: BTreeSet<PathBuf>,
    pub elm_project_root: PathBuf,
}

impl Project {
    pub fn from_dir(elm_project_root: PathBuf) -> anyhow::Result<Project> {
        // Read project elm.json
        let elm_json_str = std::fs::read_to_string(elm_project_root.join("elm.json"))
            .context("Unable to read elm.json")?;
        let config: ProjectConfig =
            serde_json::from_str(&elm_json_str).context("Invalid elm.json")?;

        let source_directories: BTreeSet<PathBuf> = match &config {
            ProjectConfig::Application(app_config) => app_config
                .source_directories
                .iter()
                .map(PathBuf::from)
                .collect(),
            ProjectConfig::Package(_) => iter::once("src".into()).collect(),
        };

        // Get absolute paths for test directories
        let mut source_or_test_directories: BTreeSet<PathBuf> = source_directories
            .iter()
            .map(|path| elm_project_root.join(path).canonicalize())
            .collect::<Result<_, _>>()?;
        // Add tests/ to the list of source directories
        if let Ok(path) = elm_project_root.join("tests").canonicalize() {
            source_or_test_directories.insert(path);
        }
        Ok(Project {
            config,
            src_and_test_dirs: source_or_test_directories,
            elm_project_root,
        })
    }

    pub fn watch(&mut self, f: impl Fn(&Self) -> anyhow::Result<()>) -> anyhow::Result<()> {
        dbg!(&self);
        // Create a channel to receive the events.
        let (tx, rx) = channel();
        // Create a watcher object, delivering debounced events.
        let mut watcher = watcher(tx, Duration::from_secs(1)).context("Failed to start watcher")?;
        let recursive = RecursiveMode::Recursive;

        let elm_json_path = self.elm_project_root.join("elm.json");

        // Watch the elm.json and the content of source directories.
        watcher
            .watch(&elm_json_path, recursive)
            .context(format!("Failed to watch {}", elm_json_path.display()))?;
        for path in &self.src_and_test_dirs {
            watcher
                .watch(path, recursive)
                .context(format!("Failed to watch {}", path.display()))?;
        }

        f(self).context("Initial run in watch mode")?;
        loop {
            match rx.recv().context("Error watching files")? {
                notify::DebouncedEvent::NoticeWrite(_) => {}
                notify::DebouncedEvent::NoticeRemove(_) => {}
                _event => {
                    // drain event queue
                    for _ in rx.try_iter() {}

                    // watch action
                    let new_project = Project::from_dir(self.elm_project_root.to_path_buf())?;
                    if self.src_and_test_dirs != new_project.src_and_test_dirs {
                        dbg!(&new_project);
                        for path in &self.src_and_test_dirs {
                            watcher
                                .unwatch(path)
                                .context(format!("Failed to unwatch {}", path.display()))?;
                        }
                        for path in &new_project.src_and_test_dirs {
                            watcher
                                .watch(path, recursive)
                                .context(format!("Failed to watch {}", path.display()))?;
                        }
                        *self = new_project;
                    }
                    f(&self).context("Subsequent run in watch mode")?;
                }
            }
        }
    }
}
