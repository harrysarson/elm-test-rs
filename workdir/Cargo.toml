[package]
name = "workdir"
version = "0.1.0"
authors = ["Harry Sarson <harry.sarson@hotmail.co.uk>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tempdir = "0.3.7"
