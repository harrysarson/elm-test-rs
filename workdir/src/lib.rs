use std::mem;
use std::path::Path;
use std::path::PathBuf;
use tempdir::TempDir;

pub enum WorkDir<P> {
    Tempory(TempDir),
    Provided(P),
}

impl<P: AsRef<Path>> AsRef<Path> for WorkDir<P> {
    #[inline]
    fn as_ref(&self) -> &Path {
        match self {
            Self::Tempory(td) => td.as_ref(),
            Self::Provided(p) => p.as_ref(),
        }
    }
}

impl<P: From<PathBuf>> WorkDir<P> {
    pub fn persist(&mut self) {
        if let Self::Tempory(_) = self {
            match mem::replace(self, Self::Provided(PathBuf::new().into())) {
                Self::Tempory(td) => *self = Self::Provided(td.into_path().into()),
                _ => {
                    panic!("impossible state");
                }
            }
        }
    }
}
