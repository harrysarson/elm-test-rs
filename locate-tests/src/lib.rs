use elm_json::project::Project;
use std::path::Path;
use tree_sitter::Tree;
use walkdir::WalkDir;
use workdir::WorkDir;

#[derive(Debug)]
pub struct ElmModule(String);

pub struct TestProjectMirror<P> {
    workdir: WorkDir<P>,
}

impl<P: AsRef<Path>> TestProjectMirror<P> {
    pub fn from_project(
        project_root: impl AsRef<Path>,
        test_path: impl AsRef<Path>,
        workdir: WorkDir<P>,
    ) -> Result<TestProjectMirror<P> {
        let test_full_path = project_root.as_ref().join(test_path);
        for entry in WalkDir::new("foo") {
            println!("{}", entry?.path().display());
        }
        // print!(
        //     "{}",
        //     sexp::parse(&tree.root_node().to_sexp()).unwrap().pretty()
        // );
        // assert!(false);
        // let root_node = tree.root_node();
        TestProjectMirror { workdir }
    }
}

#[derive(Debug)]
pub struct TestDefinitions<'a> {
    exposed: Box<[&'a str]>,
    private: Box<[&'a str]>,
}

impl<'a> TestDefinitions<'a> {
    pub fn from_sources(
        elm_json: &Project,
        project: &'a TestProjectMirror<impl AsRef<Path>>,
    ) -> TestDefinitions<'a> {
        // print!(
        //     "{}",
        //     sexp::parse(&tree.root_node().to_sexp()).unwrap().pretty()
        // );
        // assert!(false);
        // let root_node = tree.root_node();
        TestDefinitions {
            exposed: Box::new([]),
            private: Box::new([]),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    mod fixtures;

    #[test]
    fn test_test_definition_from_tree() {
        fixtures::TEST_TREES.with(|test_trees| {
            for test_tree in test_trees.iter() {
                assert_debug_snapshot!(TestDefinitions::from_sources(test_tree));
            }
        })
    }
}
