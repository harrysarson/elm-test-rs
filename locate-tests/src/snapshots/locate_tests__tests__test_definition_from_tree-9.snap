---
source: locate-tests/src/lib.rs
expression: "TestDefinitions::from_tree(test_tree)"
---
TestDefinitions {
    exposed: [],
    private: [],
}
